### ISSUE TITLE
Issues/Features for Build Shepherd `parse_metadata` Script

### ISSUE THREAD

**Issue by emmahodcroft** [nextstrain/ncov#467]
These are issues that should be tackled in order to improve the use of the `parse_metadata.py` script used by the Nextstrain build shepherds. A mix of bugs and features.

## High priority 
- [ ] **1** More specific `variants.txt` corrections. 
Example 1 : 'Gyeonggi' is a location *and* a division. We want to correct `division Gyeonggi` to `division Gyeonggi Province` but _only_ when it's a division, not when it's a location.
Example 2: 'Bahrein' is both a country and a division - both need to be corrected to 'Bahrain'. Currently only the `division` annotations are automatically generated.
Example 3: Annotations are auto-generated to correct 'unknown' to ` ` (blank). However if one has both `unknown` divisions _and_ locations, only location corrections are generated. _(Emma note: I believe these should also only go to blank when the country is 'USA' as we don't do this for other countries - right? May need to be special case)_
- [ ] **2** Auto-add missing places to the auto-generated `color_ordering.tsv` file
Example: 'Sangalkam' is missing from the `defaults/color_ordering.tsv` and is displayed with a warning (`Sangalkam (only missing in ordering)`), but this is not added to the auto-generated `developer_scripts/color_ordering.tsv`). This needs to happen before we can switch to just using an auto-generated orderings file
Note: Ensure this happens both just if the place is missing from `color_ordering` *and* if it's a 'false' division that's been corrected. See example [here](https://github.com/nextstrain/ncov/pull/455#issuecomment-657661501)

## Medium priority
- [ ] **3** Automatically add new locations needing lat-longs into `defaults/lat_longs.tsv` and re-sort the file as necessary (in order by `location`, `division`, `country`, etc)
- [ ] **4** Clarify the role of `sequences_exclude.txt` - what does it do, how can we use it better?
- [ ] **5** Clarify role of `accepted_exposure_additions.txt` and figure out how to get working optimally/resolve inconsistencies 

## Low priority
- [ ] **6** Convey to people who might want to run this script that `geopy` is necessary dependency
- [ ] **7** Clear remaining `Belgium` alerts at start of script & outline best way to handle similar alerts in future
- [ ] **8** Better handle ignoring (but informing user about) cruise-related missing data (display more compactly/prettily - and just once)

**Comment by emmahodcroft** [comment]
With regard to **1** - I wonder if we can tackle much of this by just specifying the geographic level that the replacement is happening on. Then I guess you'd need two entries if you wanted to correct *both* `country` and `division` (or `division` and `location`) but it would be more explicit.

**Comment by emmahodcroft** [comment]
PR #468 addresses part of Issue **1** (allows countries to also be modified by `variants.txt`)

**Comment by eharkins** [comment]
I'm getting 
```
Missing locations:
# USA #
California
```

but I can't find such a location: 
```
 eharkins$ cut -f 9 data/metadata.tsv | grep California
 eharkins$ 
```

Is this a known issue or possibly because of some error on my part?

**Comment by emmahodcroft** [comment]
Yes, sorry - this is due to me changing the script yesterday or day before
so it doesn't show the cruise entries anymore. However it still registers
that "California" isn't 'empty' so it prints the headers. Sorry. But I
thought this was less distracting than having the cruise ship stuff show up
over and over! But the code could be improved here, so it doesn't show if
there's only cruise ship stuff there!

**Comment by emmahodcroft** [comment]
I've updated point **8** to include this explicitly, but we could also move this to medium-priority if we'd like to tackle it sooner :)

**Comment by eharkins** [comment]
Thanks!

I can't tell - do any of these (e.g. **5**) have to do with automating creation of annotations for `additional-info-changes.txt` that comes into slack with travel history and other things to be annotated?

**Comment by emmahodcroft** [comment]
Good catch - I'll add another to the list. Unfortunately I guess the numbering system is going to get out of order now 🙃

**Comment by eharkins** [comment]
After thinking about 
>  1 More specific variants.txt corrections.
 Example 1 : 'Gyeonggi' is a location and a division. We want to correct division Gyeonggi to division Gyeonggi Province but only when it's a division, not when it's a location.

more today (this is quite common), maybe it's best to (assuming the region name and location name are spelled the same way) just always add variants which _remove_ words like province and region, and allow things like Gyeonggi to be valid regions and locations. So we would allow 
`Asia South Korea Gyeonggi Gyeonggi`
and:
`Asia South Korea Gyeonggi`
and not assert they meant one when they specified the other. 

variants would look like:
```
Gyeonggi Province -> Gyeonggi
gyeonggi -> Gyeonggi
** insert other mispellings or variants that would be applied at both division and location levels**
```

This might mean going back and correcting many examples where we already use division names including words like region and province, but doing this with this script is a lot easier.

**Comment by eharkins** [comment]
Although there will still be a need for specificity of variants I think. E.g. correcting Berkeley to Berkeley County when it is in South Carolina but not when it is in California

**Comment by emmahodcroft** [comment]
I think we've not been completely strict in the past with correcting `division` labels from 'Gyeonggi' (which is presumed to be a province/county/etc since it's a division) to 'Gyeonggi Province', so I'm happy to continue for this to be done instance by instance as long as it's consistent (shouldn't have both 'Gyeonggi' and 'Gyeonggi Province' as `division` entries). However, I don't think we need a strict rule on this, as it'll take a lot of back-correcting.

The only other thing to note here is that while it's clear to *us* that 'Gyeonggi' as a `division` will be some kind of province/county/state, this is likely *not* apparent to everyone, particularly if there's a city of the same name in the province. So, that would be an argument for being more specific in these particular cases (ex: 'Texas' is clearly not a city, but 'Gyeonggi' could be either since it's both a province and a city). 

I think to solve the first issue we can probably add something to `variants.txt` so that we specify if the correction is meant to be applied at `division` or `location` level, etc (or have two lines, if both). 

The second issue:
> Although there will still be a need for specificity of variants I think. E.g. correcting Berkeley to Berkeley County when it is in South Carolina but not when it is in California

Will need more thought, and I'm not immediately thinking of a solution. We may need to add some kind of 'condition' column? (`division==South Carolina` ?) But this is going to be fiddly....


Finally

> So we would allow
Asia South Korea Gyeonggi Gyeonggi
and:
Asia South Korea Gyeonggi
and not assert they meant one when they specified the other.

I might be misunderstanding but we should be allowing this already, right? If they specify this at both levels we take them at their word that they mean Gyeonggi as a province as division (whether or not we correct it) and Gyeonggi the city as the location. If they only give it as division, we assume it's just the province level (with no location).

**Comment by eharkins** [comment]
> I think we've not been completely strict in the past with correcting division labels from 'Gyeonggi' (which is presumed to be a province/county/etc since it's a division) to 'Gyeonggi Province', so I'm happy to continue for this to be done instance by instance as long as it's consistent (shouldn't have both 'Gyeonggi' and 'Gyeonggi Province' as division entries). However, I don't think we need a strict rule on this, as it'll take a lot of back-correcting.

Fair enough, mostly just seeking consistency moving forward for examples like this so that I have a rule of thumb to follow for cases like the [Czech regions](https://github.com/nextstrain/ncov-ingest/blob/master/source-data/location_hierarchy.tsv#L826-L847), which only sometimes use the word "Region" as part of them, and also have capital cities with the same name as the region.

> The only other thing to note here is that while it's clear to us that 'Gyeonggi' as a division will be some kind of province/county/state, this is likely not apparent to everyone, particularly if there's a city of the same name in the province. So, that would be an argument for being more specific in these particular cases (ex: 'Texas' is clearly not a city, but 'Gyeonggi' could be either since it's both a province and a city -- and to someone just looking at Auspice, they are likely to be much less clear on what a 'division' is and whether we're referring to a province or a city)

This works for me - my understanding of this is erring on the side of including words like "province" and "region", which as you point out:

> I think to solve the first issue we can probably add something to variants.txt so that we specify if the correction is meant to be applied at division or location level, etc (or have two lines, if both).

just means we might have to clarify when we want to add in those division specific words in variants.txt somehow.

> I might be misunderstanding but we should be allowing this already, right? If they specify this at both levels we take them at their word that they mean Gyeonggi as a province as division (whether or not we correct it) and Gyeonggi the city as the location. If they only give it as division, we assume it's just the province level (with no location).

You're right, this was just me thinking about ways to be consistent with including "province" and "region" as above. Before you made the good point about clarifying divisions to the auspice user, i was leaning toward excluding these words in the name of keeping variants.txt in its current simple form.

with respect to
> The second issue:
>> Although there will still be a need for specificity of variants I think. E.g. correcting Berkeley to Berkeley County when it is in South Carolina but not when it is in California

> Will need more thought, and I'm not immediately thinking of a solution. We may need to add some kind of 'condition' column? (division==South Carolina ?) But this is going to be fiddly....

this seems doable (though maybe not trivial) at first thought! In my imagination it would look like adding two columns to variants.txt so that you could specify an entry like `Berkeley   Berkeley County   division   South Carolina` and the relevant logic ([e.g.](https://github.com/nextstrain/ncov/blob/master/scripts/developer_scripts/parse_metadata.py#L377)) could be replaced with a function checking for any qualifiers (3rd and 4th column as a pair qualify the variant).

**Comment by MoiraZuber** [comment]
The latest two commits from me include the following fixes:
- Issue 1 Example 3: "Unknown" divisions are now being corrected to blank as well.
- Issue 4: This file is not needed anymore since the inconsistency was resolved - I removed all uses of the file in the code, so the file could now be deleted.
- Issue 8: I added a file called `international_exceptions.txt`, which allows for every cruise ship entry to specify where exactly we want it to be in color_ordering.tsv. This does **not** create annotations to adjust those sequences, but only adjusts the internal dataset needed for the auto-generation of the color_ordering file. The entries are temporary and can be adjusted, if another assignment seems to make more sense

**Comment by MoiraZuber** [comment]
I just discovered a bug that appears when reading in variants.txt: If there are two variants from the same original name (the case I found was `Mp -> Occitanie (Europe, France)` and `Mp -> Northern Mariana Islands (North America, USA)`), then only the second one is applied since the first is overwritten. There is a warning about that, so cases like that won't go unnoticed, but we'd need to adjust the structure of the variants dictionary to better deal with that.

**Comment by emmahodcroft** [comment]
Good spotting Moira! I'll add this so the issues on the list so we can tick it off!

**Comment by emmahodcroft** [comment]
New issues 11, 12, and 13 in 'Medium Priorities' now cover Moira's spotted bug above, the country sorting issue, and getting auto-generated credit for tweets!
